sphinx/source/docs/dev_guide/bokehjs.rst

```diff
@@ -56,6 +56,210 @@ that is compiled from several separate ``.less`` files in the BokehJS source
+for all three supported platforms have to be updated, the preferred approach is
+to generate images and compare them in CI.
+
+The full procedure for visual testing is as follows:
+
+1. Make changes to the repository and write new tests or update existing.
+2. Use ``node make tests`` to incrementally test your changes on your system.
+3. Commit changes to textual baselines (``test/baselines/*``).
+4. Push your changes to GitHub and wait for CI to finish.
+5. If you added new tests and CI will expectedly fail with "missing baseline
+   images" error message.
+6. If tests passed then your are done.
+7. If tests failed, go to bokehjs' GitHub_Actions_ page. Find the most recent
+   test run for your PR and download the associated ``bokehjs-report`` artifact.
+8. Unzip the artifact archive.
+9. Assuming devtools server is running in the background, go to ``/integration/report?platform=name``
+   where ``name`` is either ``linux``, ``macos`` or ``windows`` and review test output
+   for each platform. If there are no unintentional differences, then commit all
+   new or modified files under ``test/baselines/{linux,macos,windows}``.
+10. Push your changes again to GitHub and verify that tests pass this time.
+
+.. note::
+
+    Make sure to monitor state of ``test/baselines`` directory, so that you don't
+    commit unnecessary files. If you do so, subsequent tests will fail.
+
+Minimal model/view module
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Models (and views) come in many forms and sizes. At minimum a model is implemented.
+A view may follow if a "visual" model is being implemented. A minimal model/view
+module looks like this:
+
+.. code-block:: typescript
+
+    import {BaseModel, BaseModelView} from "models/..."
+
+    export class SomeModelView extends BaseModelView {
+      model: SomeModel
+
+      initialize(): void {
+        super.initialize()
+        // perform view intialization (remove if not needed)
+      }
+
+      async lazy_initialize(): Promise<void> {
+        await super.lazy_initialize()
+        // perform view lazy initialization (remove if not needed)
+      }
+    }
+
+    export namespace SomeModel {
+      export type Attrs = p.AttrsOf<Props>
+
+      export type Props = BaseModel.Props & {
+        some_property: p.Property<number>
+        // add more property declarations
+      }
+    }
+
+    export interface SomeModel extends SomeModel.Attrs {}
+
+    export class SomeModel extends BaseModel {
+      properties: SomeModel.Props
+
+      // do not remove this constructor, or you won't be
+      // able to use `new SomeModel({some_property: 1})`
+      constructor(attrs?: Partial<SomeModel.Attrs>) {
+        super(attrs)
+      }
+
+      static init_SomeModel(): void {
+        this.prototype.default_view = SomeModelView
+
+        this.define<SomeModel.Props>({
+          some_property: [ p.Number, 0 ],
+          // add more property definitions
+        })
+      }
+    }
```

Comment:
How can we reduce this burden for common cases? Fifty lines of boilerplate just to open the door is not only is is tedious it's also error prone since anyone would be tempted to resort to copy/paste. Do you have ideas or plans around code generation?

Reply:
This affects development of bokehjs and properly typed extensions. Extensions written in ES can skip most of the boilerplate. Simple code generation isn't possible, because you need this type information within bokehjs and outside it. Macros would be great, but typescript doesn't support those. It supports AST transforms (with access to the type system), which we already use to simplify things (well, just a bit). But writing those transforms is a lot of work. Simplifications are possible. For example the entire model namespace can be removed in 3.0, based on my recent work on the properties' system, but that will require changing all models.